train_gpt2.cu

```diff
@@ -2297,10 +2289,20 @@ void gpt2_backward(GPT2 *model) {
 }
 
 // Compute a mean of a single CPU value across all GPU processes. No-op when multi-GPU is disabled.
-float multi_gpu_cpu_float_mean(float value, const MultiGpuConfig* multi_gpu_config) {
+float multi_gpu_float_mean(float value, float *d_buffer, const MultiGpuConfig* multi_gpu_config) {
 #ifdef MULTI_GPU
-    // MPI doesn't support all reduce with mean, so we sum up, then divide.
+    if (multi_gpu_config->num_processes == 1) return value;
+
     float result;
+
+    if (multi_gpu_config->slurm_managed) {  //If the process is managed by slurm we don't need to use MPI
+        if (d_buffer == NULL) cudaCheck(cudaMalloc(&d_buffer, sizeof(float)));
+        cudaCheck(cudaMemcpy(d_buffer, &value, sizeof(float), cudaMemcpyHostToDevice));
```

Comment:
This will block the CPU execution until the copy finishes. 

It would be better to allocate a pinned CPU memory buffer (see data loader for an example), copy the CPU loss pointer to the pinned CPU buffer, then use async memcopies.

It's optional, since blocking the CPU is ~15microseconds of GPU idle time.

Reply:
Agree. Initially I was planning to do cudaMallocManaged so we don't need to keep 2 variables.
But if I use unified memory not sure I need to do prefetch before nccl function calls. Need to test this one.
If we do this via unified memory buffer implementation will be much cleaner as I feel.

Reply:
@PeterZhizhin I have introduced unified mem buffer to unblock the cpu execution and reduce the amount of variables used.